### PR TITLE
Issue #57: qbank_exportquestions

### PR DESCRIPTION
**Pull request by guillogo**

### CODE REVIEW

File: lib/classes/plugin_manager.php [hunk 1]
```diff
@@ -1940,7 +1940,7 @@ public static function standard_plugins_list($type) {
             ),
 
             'qbank' => array(
-                 ''
+                'exportquestions'
```
Comment (safatshahin) [inline lib/classes/plugin_manager.php:1943]: please add a comma at the end for further merge

File: question/export_form.php [hunk 2]
```diff
@@ -34,10 +34,15 @@
...
  * @license    http://www.gnu.org/copyleft/gpl.html GNU GPL v3 or later
+ * @deprecated since Moodle 4.0 MDL-71573
+ * @see \core_question\bank\exportquestions\export_form
  */
 class question_export_form extends moodleform {
 
     protected function definition() {
+        debugging('Class question_export_form in \core_question\export_form is deprecated,
+        please use core_question\bank\exportquestions\export_form', DEBUG_DEVELOPER);
+
```
Comment (safatshahin) [inline question/export_form.php:45]: please add " instead." at the end, just a regular moodle standard.

File: question/bank/exportquestions/classes/form/export_form.php [hunk 3]
```diff
@@ -0,0 +1,93 @@
...
+/**
+ * Defines the export questions form.
+ *
+ * @package    moodlecore
+ * @subpackage questionbank
+ * @copyright  2007 Jamie Pratt me@jamiep.org
+ * @license    http://www.gnu.org/copyleft/gpl.html GNU GPL v3 or later
+ */
+
+
```
Comment (safatshahin) [inline question/bank/exportquestions/classes/form/export_form.php:26]: please add namespace qbank_exportquestions\form

File: question/bank/exportquestions/classes/form/export_form.php [hunk 4]
```diff
@@ -0,0 +1,93 @@
...
+// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+// GNU General Public License for more details.
+//
+// You should have received a copy of the GNU General Public License
+// along with Moodle.  If not, see <http://www.gnu.org/licenses/>.
+
+/**
+ * Defines the export questions form.
+ *
+ * @package    moodlecore
```
Comment (safatshahin) [inline question/bank/exportquestions/classes/form/export_form.php:20]: please use the package qbank_exportquestions, there is no subpackage

File: question/export_form.php [hunk 5]
```diff
@@ -34,10 +34,15 @@
  *
  * @copyright  2007 Jamie Pratt me@jamiep.org
  * @license    http://www.gnu.org/copyleft/gpl.html GNU GPL v3 or later
+ * @deprecated since Moodle 4.0 MDL-71573
+ * @see \core_question\bank\exportquestions\export_form
```
Comment (safatshahin) [inline question/export_form.php:38]: please use the namespace qbank_exportquestions\form

**Comment by safatshahin** [comment]
Hi @guillogo 
I have made some comments, in addition to those, there are a couple of method which are exclusive to export plugin, eg question_default_export_filename, question_make_export_url, these can come to the plugin and the existing ones in questionlib can be deprecated.

**Comment by safatshahin** [comment]
Hi @guillogo 
Please also check lib/classes/event/question_exported.php
the url need to be changed.

File: question/bank/exportquestions/export.php [hunk 6]
```diff
@@ -24,13 +24,14 @@
...
+require_once(__DIR__ . '/../../../config.php');
 require_once($CFG->dirroot . '/question/editlib.php');
-require_once($CFG->dirroot . '/question/export_form.php');
+require_once($CFG->dirroot . '/question/bank/exportquestions/classes/form/export_form.php');
 require_once($CFG->dirroot . '/question/format.php');
+require_once($CFG->dirroot . '/question/renderer.php');
 
 list($thispageurl, $contexts, $cmid, $cm, $module, $pagevars) =
-        question_edit_setup('export', '/question/export.php');
+        question_edit_setup('export', '/question/bank/exportquestions/export.php');
```
Comment (safatshahin) [inline question/bank/exportquestions/export.php:34]: please change the url in the event as well.

**Comment by guillogo** [comment]
Thank you @safatshahin for looking into this, the changes were done. Thank you.

File: question/bank/exportquestions/classes/exportquestions_helper.php [hunk 7]
```diff
@@ -0,0 +1,58 @@
...
+//
+// You should have received a copy of the GNU General Public License
+// along with Moodle.  If not, see <http://www.gnu.org/licenses/>.
+
+/**
+ * Library functions used by question/preview.php.
+ *
+ * @package    qbank_exportquestions
+ * @copyright  2010 The Open University
+ * @author     2021 Safat Shahin <safatshahin@catalyst-au.net>
```
Comment (safatshahin) [inline question/bank/exportquestions/classes/exportquestions_helper.php:22]: please change to your name and if the code is from questionlib, copyright should follow that.

**Comment by safatshahin** [comment]
thanks @guillogo 
almost done, just a little change in the copyright as its your code, should have your name.

File: question/bank/exportquestions/classes/exportquestions_helper.php (hunk 7 above)
Comment (safatshahin) [reply question/bank/exportquestions/classes/exportquestions_helper.php:22]: maybe something like this

/**
 * Helper class for adding/editing a question.
 * 
 * This code is based on question/editlib.php by Martin Dougiamas.
 *
 * @package    qbank_editquestion
 * @copyright  2021 Safat Shahin <safatshahin@catalyst-au.net>   -- your name
 * @license    http://www.gnu.org/copyleft/gpl.html GNU GPL v3 or later
 */

Comment (guillogo) [reply question/bank/exportquestions/classes/exportquestions_helper.php:22]: You're right I missed that, thank you.

**Comment by guillogo** [comment]
Thank you @safatshahin done now.

File: question/bank/exportquestions/classes/exportquestions_helper.php [hunk 8]
```diff
@@ -0,0 +1,57 @@
...
+// GNU General Public License for more details.
+//
+// You should have received a copy of the GNU General Public License
+// along with Moodle.  If not, see <http://www.gnu.org/licenses/>.
+
+/**
+ * Library functions used by qbank_exportquestions/export.php.
+ *
+ * @package    qbank_exportquestions
+ * @copyright 2021, Guillermo Gomez Arias <guillermogomez@catalyst-au.net>
```
Comment (safatshahin) [inline question/bank/exportquestions/classes/exportquestions_helper.php:21]: one more space please/
also can you please add the info of which code its based on.

**Comment by safatshahin** [comment]
Hi @guillogo 
i have one tiny comment. as a part of this, can you please add the privacy api in this PR?

**Comment by guillogo** [comment]
Hi @safatshahin 

Done:
1. Branch rebased.
2. Provider class added.
3. Fix copyright comment.

Thank you.

File: question/bank/exportquestions/classes/exportquestions_helper.php [hunk 9]
```diff
@@ -0,0 +1,59 @@
...
+// You should have received a copy of the GNU General Public License
+// along with Moodle.  If not, see <http://www.gnu.org/licenses/>.
+
+/**
+ * Library functions used by qbank_exportquestions/export.php.
+ *
+ * This code is based on lib/questionlib.php by Martin Dougiamas.
+ *
+ * @package    qbank_exportquestions
+ * @copyright  2021, Guillermo Gomez Arias <guillermogomez@catalyst-au.net>
```
Comment (safatshahin) [inline question/bank/exportquestions/classes/exportquestions_helper.php:23]: please change the copyright according to dan's comment

File: question/bank/exportquestions/classes/privacy/provider.php [hunk 10]
```diff
@@ -0,0 +1,37 @@
...
+// (at your option) any later version.
+//
+// Moodle is distributed in the hope that it will be useful,
+// but WITHOUT ANY WARRANTY; without even the implied warranty of
+// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+// GNU General Public License for more details.
+//
+// You should have received a copy of the GNU General Public License
+// along with Moodle.  If not, see <http://www.gnu.org/licenses/>.
+
```
Comment (safatshahin) [inline question/bank/exportquestions/classes/privacy/provider.php:16]: please add copyright here

File: question/bank/exportquestions/classes/privacy/provider.php [hunk 11]
```diff
@@ -0,0 +1,37 @@
...
+namespace qbank_exportquestions\privacy;
+
+/**
+ * Privacy Subsystem for qbank_exportquestions implementing null_provider.
+ *
+ * @package   qbank_exportquestions
+ * @category  privacy
+ * @copyright 2021, Guillermo Gomez Arias <guillermogomez@catalyst-ca.net>
```
Comment (safatshahin) [inline question/bank/exportquestions/classes/privacy/provider.php:24]: please change the copyright according to dan's comment

File: question/bank/exportquestions/export.php [hunk 12]
```diff
@@ -24,21 +24,27 @@
  */
 
 
-require_once(__DIR__ . '/../config.php');
+require_once(__DIR__ . '/../../../config.php');
 require_once($CFG->dirroot . '/question/editlib.php');
-require_once($CFG->dirroot . '/question/export_form.php');
+require_once($CFG->dirroot . '/question/bank/exportquestions/classes/form/export_form.php');
```
Comment (safatshahin) [inline question/bank/exportquestions/export.php:29]: probably dont need this as you are using namespace, please have a look

File: question/bank/exportquestions/export.php [hunk 13]
```diff
@@ -24,21 +24,27 @@
  */
```
Comment (safatshahin) [inline question/bank/exportquestions/export.php:24]: please fix the package

File: question/bank/exportquestions/lang/en/qbank_exportquestions.php [hunk 14]
```diff
@@ -0,0 +1,26 @@
...
+// GNU General Public License for more details.
+//
+// You should have received a copy of the GNU General Public License
+// along with Moodle.  If not, see <http://www.gnu.org/licenses/>.
+
+/**
+ * Strings for component 'qbank_exportquestions', language 'en'
+ *
+ * @package   qbank_exportquestions
+ * @copyright 2021, Guillermo Gomez Arias <guillermogomez@catalyst-au.net>
```
Comment (safatshahin) [inline question/bank/exportquestions/lang/en/qbank_exportquestions.php:21]: please change the copyright according to dan's comment

File: question/bank/exportquestions/version.php [hunk 15]
```diff
@@ -0,0 +1,30 @@
...
+// GNU General Public License for more details.
+//
+// You should have received a copy of the GNU General Public License
+// along with Moodle.  If not, see <https://www.gnu.org/licenses/>.
+
+/**
+ * Plugin version and other meta-data are defined here.
+ *
+ * @package   qbank_exportquestions
+ * @copyright 2021, Guillermo Gomez Arias <guillermogomez@catalyst-au.net>
```
Comment (safatshahin) [inline question/bank/exportquestions/version.php:21]: please change the copyright according to dan's comment

**Comment by safatshahin** [comment]
Hi @guillogo 
Just some quick changes required with the comments. 
Please rebase it against the latest base branches to get the qbank enable and disable feature.
An important point to implement is, making changes to plugin to disable features if the plugin is disabled.
thanks

**Pull request closed without merging**